### PR TITLE
JDK-8320207: doclet incorrectly chooses code font for a See Also link

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a change to the font used to display the content of an `@see` tag.

The underlying trigger for this is in `ClassFileFormatVersion.java` which contains 
  `@see System#getProperties System property {@code java.class.version}`

The (reasonable) presumption is that by default the label will be in plain font, except for the trailing `{@code...}` tag.
Because `SeeTaglet` always generates a link with code font, the nested tag gives a nested `<code>...</code>` element, which is reported as an error by `html-tidy`.

However, there are numerous examples in the JDK report where there is a less-reasonable presumption that the output _will_ be in code font.  This is often used to give a "slightly different" rendering of the target, such as omitting the parentheses and any parameters from a link to a method.  There are too many such examples to easily change, even though it would be better to do so for consistency.

The dilemma is resolved in favor of not using code font when the label looks like a phrase and not a form of the target signature, on the grounds that it is better to omit `<code>...</code>` and allow the author to opt-in to using code font either explicitly or with a `{@code tag}`, since there is no way of opt-out of being in code font within the content of the element.

Thus, the fix is a change to `SeeTaglet` which analyses the label to see if it appears to be a phrase of some sort, and not a form of a reference to the target. If it appears to be a phrase, code font is not used; if it appears to be a reference to the target, code font is used.

Note: initially, the solution was more focussed on examining the label in more detail and matching it more accurately with the reference, but there are enough variations in the JDK code that this was deemed impractical. The solution to focus on whether the label looks like a phrase is simpler and more reliable.

This does change the font used for some links in the JDK documentation, including the link in `ClassFileFormatVersion` described above.  (Only the font, as indicated by the use of `<code>` is changed, never the target URL or the text of the link.) Many of the other changes are as expected, such as links to "security properties" or "Types and Elements".  There are a few places where the label text is a mixture of a signature and an explanation, which used to be all in code font and is now all in plain font. For these cases, it would be good to (separately) modify the label to explicitly use code font around the relevant parts of the label.  As an example of these "mixed use" tags, see the following in `com.sun.jdi.Type`
```
 * @see PrimitiveType Subinterface PrimitiveType
 * @see ReferenceType Subinterface ReferenceType
 * @see Value Value - for relationship between Type and Value
 * @see Field#type() Field.type() - for usage examples
 ```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8320337](https://bugs.openjdk.org/browse/JDK-8320337) to be approved

### Issues
 * [JDK-8320207](https://bugs.openjdk.org/browse/JDK-8320207): doclet incorrectly chooses code font for a See Also link (**Bug** - P3)
 * [JDK-8320337](https://bugs.openjdk.org/browse/JDK-8320337): doclet incorrectly chooses code font for a See Also link (**CSR**)

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.org/census#hannesw) (@hns - **Reviewer**) ⚠️ Review applies to [75aeb6c0](https://git.openjdk.org/jdk/pull/16699/files/75aeb6c05e76ed8931059e350ca014d044096cc5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16699/head:pull/16699` \
`$ git checkout pull/16699`

Update a local copy of the PR: \
`$ git checkout pull/16699` \
`$ git pull https://git.openjdk.org/jdk.git pull/16699/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16699`

View PR using the GUI difftool: \
`$ git pr show -t 16699`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16699.diff">https://git.openjdk.org/jdk/pull/16699.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16699#issuecomment-1815399262)

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/SeeTaglet.java [hunk 1]
```diff
@@ -189,7 +190,124 @@ private Content seeTagOutput(Element element, SeeTree seeTag) {
 
             default -> throw new IllegalStateException(ref0.getKind().toString());
         }
+    }
+
+    /**
+     * {@return {@code true} if the label should be rendered in plain font}
+     *
+     * The method uses a heuristic, to see string form of the label is a substring of the reference.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/SeeTaglet.java:198]:
```suggestion
     * The method uses a heuristic, to see if the string form of the label is a substring of the reference.
```

**Comment by pavelrappo** [comment]
> The underlying trigger for this is in `ClassFileFormatVersion.java` which contains `@see System#getProperties System property {@code java.class.version}`
> 

Separately and just to double-check: is it supposed to be like that or should that site use `{@systemProperty}` instead? Perhaps this question should be equally addressed to @jddarcy, the author of the respective [change](https://github.com/openjdk/jdk/pull/9299).

**Review by hns — APPROVED** [review]
Looks good to me. Although this seems like an intricate solution for what at first sight looks like a simple issue, the actual occurrences in our code base show there's an actual problem to be solved. 

Although this looks like something that could "just work", like Pavel I wonder if this should be documented in the spec.

**Comment by jonathan-gibbons** [comment]
> > The underlying trigger for this is in `ClassFileFormatVersion.java` which contains `@see System#getProperties System property {@code java.class.version}`
> 
> Separately and just to double-check: is it supposed to be like that or should that site use `{@systemProperty}` instead? Perhaps this question should be equally addressed to @jddarcy, the author of the respective [change](https://github.com/openjdk/jdk/pull/9299).

Yes, that's a separate question, which does not affect the overall issue that was uncovered as a result, and addressed here.

File: test/langtools/jdk/javadoc/doclet/testSeeTag/TestSeeTagFont.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2002, 2023, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testSeeTag/TestSeeTagFont.java:2]: Fixed.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/SeeTaglet.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/SeeTaglet.java:198]: thanks; fixed

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/SeeTaglet.java [hunk 3]
```diff
@@ -189,7 +190,124 @@ private Content seeTagOutput(Element element, SeeTree seeTag) {
...
+     * <li>There are nested parentheses, or content after the parentheses,
+     *     which cannot occur in a standalone signature
+     * <li>The simple name inferred from the reference does not match
+     *     any simple name inferred from the label
+     * </ul>
+     *
+     * @param refSignature the signature of the target of the reference
+     * @param label the label
+     */
+    private boolean isPlain(String refSignature, List<? extends DocTree> label) {
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/SeeTaglet.java:223]: I'm writing a CSR.

**Pull request closed without merging**